File: Question_section_Semantic_Search/Python_Ver_Question_semantics/src/semantic_search/mongo_store.py
from datetime import datetime
from typing import List, Dict
from pymongo import MongoClient

class MongoVectorStore:
    def __init__(self, uri: str, db_name: str, coll_name: str):
        self.client = MongoClient(uri)
        self.coll = self.client[db_name][coll_name]

    def upsert_many(self, docs: List[Dict], embeddings: List[List[float]]):
        for doc, emb in zip(docs, embeddings):
            payload = dict(doc)
            payload["embedding"] = emb
            payload["created_at"] = datetime.utcnow()
            self.coll.update_one({"id": payload["id"]}, {"$set": payload}, upsert=True)

    def vector_search(self, query_vec: List[float], index_name: str, limit: int = 1, num_candidates: int = 100):
        pipeline = [
            {
                "$vectorSearch": {
                    "index": index_name,
                    "path": "embedding",
                    "queryVector": query_vec,
                    "numCandidates": num_candidates,
                    "limit": limit
                }
            },
            {
                "$project": {
                    "_id": 0,
                    "id": 1,
                    "question": 1,
                    "answer": 1,
                    "section_number": 1,
                    "section_title": 1,
                    "created_at": 1,
                    "score": {"$meta": "vectorSearchScore"}
                }
            }
        ]
        return list(self.coll.aggregate(pipeline))

    def create_ttl_index(self, days: int = 7):
        self.coll.create_index("created_at", expireAfterSeconds=days * 24 * 3600)

File: Question_section_Semantic_Search/NodeJs_Ver_Question_semantics/src/mongoStore.js
import { MongoClient } from "mongodb";

export class MongoVectorStore {
  constructor(uri, dbName, collName) {
    this.client = new MongoClient(uri);
    this.dbName = dbName;
    this.collName = collName;
    this.collection = null;
  }

  async init() {
    await this.client.connect();
    this.collection = this.client.db(this.dbName).collection(this.collName);
  }

  async upsertMany(docs, embeddings) {
    for (let i = 0; i < docs.length; i++) {
      const payload = {
        ...docs[i],
        embedding: embeddings[i],
        created_at: new Date()
      };
      await this.collection.updateOne(
        { id: payload.id },
        { $set: payload },
        { upsert: true }
      );
    }
  }

  async vectorSearch(queryVec, indexName, limit = 1, numCandidates = 100) {
    const pipeline = [
      {
        $vectorSearch: {
          index: indexName,
          path: "embedding",
          queryVector: queryVec,
          numCandidates,
          limit
        }
      },
      {
        $project: {
          _id: 0,
          id: 1,
          question: 1,
          answer: 1,
          section_number: 1,
          section_title: 1,
          created_at: 1,
          score: { $meta: "vectorSearchScore" }
        }
      }
    ];
    return await this.collection.aggregate(pipeline).toArray();
  }

  async createTTLIndex(days = 7) {
    await this.collection.createIndex(
      { created_at: 1 },
      { expireAfterSeconds: days * 24 * 3600 }
    );
  }

  async createTextIndex() {
    await this.collection.createIndex({ question: "text" });
  }

  async createVectorSearchIndex(indexName, dimensions = 384) {
    const indexDef = {
      name: indexName,
      type: "vectorSearch",
      definition: {
        fields: [
          {
            path: "embedding",
            type: "vector",
            numDimensions: dimensions,
            similarity: "cosine"
          }
        ]
      }
    };

    await this.collection.createSearchIndex(indexDef);
  }
}